discovery.go

```diff
@@ -0,0 +1,140 @@
+package viamrtsp
+
+import (
+	"bytes"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"go.viam.com/rdk/logging"
+)
+
+const (
+	maxUDPMessageBytesSize     = 1536
+	standardWSDiscoveryAddress = "239.255.255.250:3702"
+	discoveryTimeout           = 10 * time.Second
+)
+
+// RTSPDiscovery is responsible for discovering RTSP camera devices using WS-Discovery and ONVIF.
+type RTSPDiscovery struct {
+	multicastAddress string
+	logger           logging.Logger
+	conn             *net.UDPConn
+}
+
+// newRTSPDiscovery creates a new RTSPDiscovery instance with default values.
+func newRTSPDiscovery(logger logging.Logger) *RTSPDiscovery {
+	return &RTSPDiscovery{
+		multicastAddress: standardWSDiscoveryAddress,
+		logger:           logger,
+	}
+}
+
+// Close closes the UDP connection if it exists.
+func (d *RTSPDiscovery) close() error {
+	if d.conn != nil {
+		err := d.conn.Close()
+		d.conn = nil
+		return err
+	}
+	return nil
+}
+
+// generateDiscoveryMessage formats an xml discovery message properly.
+func (d *RTSPDiscovery) generateDiscoveryMessage() string {
+	messageID := uuid.New().String()
+	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
+	<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope" 
+                xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" 
+                xmlns:wsdd="http://schemas.xmlsoap.org/ws/2005/04/discovery">
+		<SOAP-ENV:Header>
+			<wsa:MessageID>uuid:%s</wsa:MessageID>
+			<wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
+			<wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</wsa:Action>
+		</SOAP-ENV:Header>
+		<SOAP-ENV:Body>
+			<wsdd:Probe>
+				<wsdd:Types>dn:NetworkVideoTransmitter</wsdd:Types>
+			</wsdd:Probe>
+		</SOAP-ENV:Body>
+	</SOAP-ENV:Envelope>`, messageID)
+}
+
+// discoverRTSPAddresses performs a WS-Discovery and extracts RTSP addresses from the XAddrs field.
+func (d *RTSPDiscovery) discoverRTSPAddresses() ([]string, error) {
+	var discoveredAddresses []string
+
+	addr, err := net.ResolveUDPAddr("udp4", d.multicastAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve UDP address: %w", err)
+	}
+
+	d.conn, err = net.ListenUDP("udp4", nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create UDP socket: %w", err)
+	}
+	defer d.close()
+
+	_, err = d.conn.WriteToUDP([]byte(d.generateDiscoveryMessage()), addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send discovery message: %w", err)
+	}
+
+	err = d.conn.SetReadDeadline(time.Now().Add(discoveryTimeout))
+	if err != nil {
+		return nil, fmt.Errorf("failed to set read deadline: %w", err)
+	}
```

Comment:
This pattern makes the code slightly more compact in go
```suggestion
	if err = d.conn.SetReadDeadline(time.Now().Add(discoveryTimeout)); err != 
		return nil, fmt.Errorf("failed to set read deadline: %w", err)
	}
```